### PR TITLE
DiagnosticsMixin

### PR DESCRIPTION
**Pull request by Yuriyzabegaev**
## Proposed changes

Diagnostics mixin. Currently supports dividing the matrix into blocks based on different equations and subdomains/interfaces. Can plot:
    - condition number of each block
    - absolute maximum value of each block
Supports gathering data about one equation/variable block from all subdomains.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply_

- [ ] Minor change (e.g., dependency bumps, broken links, etc).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code, etc).
- [ ] Other: 

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.	
- [x] Static typing is included in the update.
- [x] This PR does not duplicated existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).

### CODE REVIEW

File: src/porepy/viz/diagnostics_mixin.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        It is assumed that the full Jacobian matrix is stored in `self.linear_system`,
+        and full information about the matrix indices is stored in
+        `self.equation_system._equation_image_space_composition`.
+
+        Note:
+            It is assumed that variables with the same name defined on different grids
+            are semantically the same variable.
+
+        Args:
+            grouping (optional): Either "subdomains" or None. If "subdomains",
```
Comment (keileg) [inline src/porepy/viz/diagnostics_mixin.py:84]: How difficult will it be to allow for computing for some subdomains only? I think this would require a second argument (Optional[list[pp.Grid]], None gives all subdomains, empty list give no subdomain). I think this could be useful, what do you say?

File: src/porepy/viz/diagnostics_mixin.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        block_data: dict[tuple[int, int], dict] = defaultdict(dict)
+        for i, j in product(range(block_matrix_shape[0]), range(block_matrix_shape[1])):
+            submat = self.submatrices[i, j]
+            is_empty_block = submat.data[submat.data != 0].size == 0
+            block_data[i, j]["is_empty_block"] = is_empty_block
+            if not is_empty_block:
+                if is_plotting_condition_number:
+                    block_data[i, j]["condition_number"] = get_condition_number(submat)
+                if is_plotting_max:
+                    block_data[i, j]["max"] = abs(submat.data).max()
```
Comment (keileg) [inline src/porepy/viz/diagnostics_mixin.py:152]: I think you can expand the utility of this function by adding kwargs on the form dict[str, Callable[[Matrix], float]] and then at this point do something like

for key, func in kwargs.items()
    block_data[i, j][key] = func(submat)

If you also return block_data, the user can have increased flexibility in doing tailored diagnostics.

File: src/porepy/viz/diagnostics_mixin.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+    def _equations_indices(
+        self, grouping: Optional[Literal["subdomains"]] = None
+    ) -> dict[str, np.ndarray]:
+        """Collects the indices of equations presented in the Jacobian matrix as rows.
+
+        Returns:
+            A dictionary with keys - printed names of the equations, values - indices of
+            this equation in the global Jacobian matrix.
+        """
```
Comment (keileg) [inline src/porepy/viz/diagnostics_mixin.py:179]:
```suggestion
   
        """
```

File: src/porepy/viz/diagnostics_mixin.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+        equation_indices = {}
+        assert grouping in (None, "subdomains"), f"Unknown grouping: {grouping}"
+
+        assembled_equation_indices = self.equation_system.assembled_equation_indices
+
+        if grouping is None:
+            block_indices = 0
+            for eq_name, dof_indices in assembled_equation_indices.items():
+                equation_image_space = (
```
Comment (keileg) [inline src/porepy/viz/diagnostics_mixin.py:188]: I am struggling to see the ordering of the dof indices for a random subdomain here. There is no orderign in the loop, which means the order will be the same as in equation_image_space (since dictionaries preserve order). I guess this is okay, but have you thought through this?

File: src/porepy/viz/diagnostics_mixin.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+        assert grouping in (None, "subdomains"), f"Unknown grouping: {grouping}"
+
+        if grouping is None:
+            for variable in self.equation_system.variables:
+                grid_variable_name = _format_variable_name_with_grid(variable)
+                dofs = self.equation_system.dofs_of([variable])
+                variable_indices[grid_variable_name] = dofs
+
+        elif grouping == "subdomains":
+            previous_variable_name = None
```
Comment (keileg) [inline src/porepy/viz/diagnostics_mixin.py:232]: Why is this necessary? Please comment.

File: src/porepy/viz/diagnostics_mixin.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+            equation_indices: The dictionary with keys - written names of equations and
+                values - indices of these equations in the Jacobian. (Rows index).
+            variable_indices: The dictionary with keys - written names of variables and
+                values - indices of these variables in the Jacobian (Columns index).
+        """
+        submatrices = {}
+
+        for i, ind_row in enumerate(equation_indices.values()):
+            for j, ind_col in enumerate(variable_indices.values()):
+                submatrix_indices = np.meshgrid(
```
Comment (keileg) [inline src/porepy/viz/diagnostics_mixin.py:263]: Why not use slices?

File: src/porepy/viz/diagnostics_mixin.py [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+                    ind_row, ind_col, sparse=True, indexing="ij"
+                )
+                submatrices[i, j] = mat[tuple(submatrix_indices)]
+
+        self.submatrices = submatrices
```
Comment (keileg) [inline src/porepy/viz/diagnostics_mixin.py:268]: Is there any reason not to return this, rather than storing it? I am afraid this will effectively lead to a double storage of the matrix.

File: tests/integration/test_diagnostics_mixin.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
+""" Tests of functionality of porepy.vis.diagnostics_mixin.DiagnosticsMixin.
```
Comment (keileg) [inline tests/integration/test_diagnostics_mixin.py:1]: Placement of this file: The other model-related mixins are located in tests/integration/models. I suggest moving this as well.

File: tests/integration/test_diagnostics_mixin.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+# It changes plt.show to a mockup that does nothing.
+# Thus, we prevent blocking of plt.show.
+# This provides first argument to the function, which we don't use.
+def test_diagnostics_mixin(_) -> None:
+    """Runs over the functionality of
+    :class:`~porepy.viz.diagnostics_mixin.DiagnosticsMixin`.
+
+    Plotting functionality is based on seaborn which might not be installed in the
+    testing environment. In this case the mixin must fall back to text output mode,
+    and test must run succeed. Thus, we test everything except plotting.
```
Comment (keileg) [inline tests/integration/test_diagnostics_mixin.py:25]: You mean test must run successfully?

File: tests/integration/test_diagnostics_mixin.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    setup.show_diagnostics(
+        is_plotting_condition_number=True, is_plotting_max=True, grouping=None
+    )
+    setup.show_diagnostics(
+        is_plotting_condition_number=True, is_plotting_max=True, grouping="subdomains"
+    )
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (keileg) [inline tests/integration/test_diagnostics_mixin.py:46]: I don't think this is necesary since we run tests with pytest, which has its own way of doing test discovery.

File: src/porepy/viz/diagnostics_mixin.py [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+    Index represents the pair (row, column) of the block.
+    """
+
+    # The mixin expects the model to have these properties defined.
+    linear_system: tuple[csr_matrix, np.ndarray]
+    equation_system: EquationSystem
+
+    def show_diagnostics(
+        self,
+        grouping: Optional[Literal["subdomains"]] = None,
```
Comment (keileg) [inline src/porepy/viz/diagnostics_mixin.py:68]: How about interface variables and equations?

**Review by keileg — COMMENTED** [review]
Suggestions and comment. Additional remarks on the tutorial to be posted soon.

File: tutorials/diagnostics.ipynb [hunk 12]
```diff
@@ -0,0 +1,392 @@
+{
```
Comment (keileg) [inline tutorials/diagnostics.ipynb:1]: <p>I suggest making a remark that you solve the problem and plot the solution.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/813/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (keileg) [inline tutorials/diagnostics.ipynb:1]: <p>You really do two things here: Access the Jacobian and plot the sparsity pattern. In a tutorial you want to be explicit on such nuances.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/813/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (keileg) [inline tutorials/diagnostics.ipynb:1]: <p>I like this.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/813/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (keileg) [inline tutorials/diagnostics.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.    setup = MandelWithDiagnostics({"domain_size": (10, 10)})
</pre><p>Why change the model in the middle of the tutorial?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/813/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (keileg) [inline tutorials/diagnostics.ipynb:1]: <p>This figure corresponds to a mass and energy balance equation, not Mandel.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pmgbergen/porepy/pull/813/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

File: src/porepy/viz/diagnostics_mixin.py [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+
+        def get_eig(mat, which) -> float:
+            return svds(mat, k=1, return_singular_vectors=False, which=which).item()
+
+        def get_condition_number(mat) -> float:
+            largest_eig = get_eig(mat, which="LM")
+            # Smallest eigenvalue might be zero.
+            smallest_eig = get_eig(mat, which="SM")
+            if mat.data.size == 0:
+                return 0.0
```
[the remark this answers is not in the document]
Comment (Yuriyzabegaev) [reply src/porepy/viz/diagnostics_mixin.py:106]: Right now, it's not clear how to distinguish between these cases.

File: src/porepy/viz/diagnostics_mixin.py (hunk 6 above)
Comment (Yuriyzabegaev) [reply src/porepy/viz/diagnostics_mixin.py:263]: After I implemented custom grouping, it might be the case that dofs are not contiguous. For example, what if we have three subdomains, and want to group 0th and 2nd in one block? Meshgrid uses a "sparse" argument, so it should not use too much memory.

File: src/porepy/viz/diagnostics_mixin.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/viz/diagnostics_mixin.py:107]: Fair enough.

File: src/porepy/viz/diagnostics_mixin.py (hunk 6 above)
Comment (keileg) [reply src/porepy/viz/diagnostics_mixin.py:265]: Makes sense.

File: src/porepy/viz/diagnostics_mixin.py [hunk 14]
```diff
@@ -0,0 +1,486 @@
...
+from porepy.grids.md_grid import MixedDimensionalGrid
+from porepy.grids.mortar_grid import MortarGrid
+from porepy.numerics.ad.equation_system import EquationSystem
+from porepy.numerics.ad.operators import Variable
+
+# Seaborn is a visualization library based on Matplotlib. It allows for building nice
+# figures. Seaborn library is not one of the dependencies of PorePy. Thus, it might not
+# be present on the user's device. In this case, Python raises ImportError exception.
+try:
+    # PyLint is not happy with Seaborn since it's not typed. We silence this warning.
```
Comment (keileg) [inline src/porepy/viz/diagnostics_mixin.py:30]: You mean mypy?

File: src/porepy/viz/diagnostics_mixin.py [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+                is big. Defaults to False.
+            is_plotting_max (optional): Whether to plot the absolute maximum.
+                This option is computationally cheap even for big matrices.
+            additional_handlers (optional): A dictionary of additional functions to be
+                applied to the submatrices. The keys represent the name of the handler.
+                The values are the functions with the arguments: the matrix, the
+                equation name and the variable name. The equation and the variable names
+                are passed to allow the user for calculating the handler value only in a
+                subset of equations and variables of interest.
+        Returns:
```
Comment (keileg) [inline src/porepy/viz/diagnostics_mixin.py:127]:
```suggestion

        Returns:
``````suggestion
        Returns:
```

**Review by keileg — APPROVED** [review]
Two minor comments. Feel free to merge when you have fixed them.
